### PR TITLE
Feature/ida6 1317 migrate user profile end point to accounts users api service

### PR DESCRIPTION
**Pull request by griffithsjd**
__Short description outlining key changes/additions__
Ported code across from authentication-service
Add /user/profile to load balancer

__JIRA Ticket Number__
IDVA6-1317

### Type of change

* [ ] Bug fix
* [x] New feature
* [ ] Breaking change
* [ ] Infrastructure change

### Pull request checklist

* [x] I have added unit tests for new code that I have added
* [x] I have added/updated functional tests where appropriate
* [x] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__

### CODE REVIEW

**Review by modhasanjay — APPROVED** [review]

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/accounts/user/controller/GetUserRecordController.java [hunk 1]
```diff
@@ -40,4 +60,40 @@ public ResponseEntity<User> getUserDetails( final String userId, final String xR
...
+                return ResponseEntity.status(HttpServletResponse.SC_NOT_FOUND).body(new HashMap<>(Map.of("error", "Cannot locate user")));
+            }
+            final var user = userOptional.get();
+
+            userProfile.put(FORENAME, userDetails.getForename());
+            userProfile.put(SURNAME, userDetails.getSurname());
+            userProfile.put(EMAIL, userDetails.getEmail());
+            userProfile.put(ID, userDetails.getUserID());
+            userProfile.put(PRIVATE_BETA_USER, user.getIsPrivateBetaUser());
+            userProfile.put(ONE_LOGIN_DATA, user.getHasLinkedOneLogin());
```
Comment (jackthurkettle) [inline src/main/java/uk/gov/companieshouse/accounts/user/controller/GetUserRecordController.java:88]: This should be "account_type" and have a string value

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/accounts/user/controller/GetUserRecordController.java [hunk 2]
```diff
@@ -40,4 +59,39 @@ public ResponseEntity<User> getUserDetails( final String userId, final String xR
         return new ResponseEntity<>( user, HttpStatus.OK );
     }
 
+    @GetMapping("/user/profile")
+    public ResponseEntity<Object> getUserProfile(HttpServletRequest request) {
+        Map<String,Object> userProfile = new HashMap<>();
+        try {
+            // UserProfilePermissionInterceptor attaches to request
```
Comment (jackthurkettle) [inline src/main/java/uk/gov/companieshouse/accounts/user/controller/GetUserRecordController.java:66]: Interceptor is called "Oauth2AuthorizationInteceptor"

File: src/main/java/uk/gov/companieshouse/accounts/user/interceptor/Oauth2AuthorizationInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        if (oad.getTokenValidUntil() < LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)) {
+            logger.traceRequest(request, "Token expired", null);
+            response.addHeader(WWW_AUTHENTICATE, "Bearer error='expired_token' error-desc='Access token has expired' realm='https://account.companieshouse.gov.uk'");
+            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, INVALID_AUTHORIZATION_HEADER);
+            return false;
+        }
+
+        request.setAttribute(ACCESS_CONTROL_ALLOW_ORIGIN, request.getHeaders(HttpHeaders.ORIGIN));
+
+        //All checks passed, stash oauth2_authorisation and allow request to continue
```
Comment (jackthurkettle) [inline src/main/java/uk/gov/companieshouse/accounts/user/interceptor/Oauth2AuthorizationInterceptor.java:139]: Not using stash, attaching to request instead

File: pom.xml [hunk 4]
```diff
@@ -135,6 +135,11 @@
             <artifactId>mapstruct</artifactId>
             <version>${org.mapstruct.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20231013</version>
```
Comment (emead) [inline pom.xml:141]: Is there a reason to not use the latest version 20240303?

**Comment by emead** [comment]
See code smells and test coverage in SonarQube report: https://code-analysis.platform.aws.chdev.org/dashboard?id=uk.gov.companieshouse%3Aaccounts-user-api&pullRequest=51

File: src/main/java/uk/gov/companieshouse/accounts/user/interceptor/Oauth2AuthorizationInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+        String[] splitHeader = authHeader.split("\\s");
+        var authType = splitHeader[0];
+        String authValue;
+        if (authType.equalsIgnoreCase("basic")) {
+            authValue = new String(Base64.getDecoder().decode(splitHeader[1]));
+
+            if (!authValue.endsWith(":")) {
+                logger.trace("Invalid username:password string");
+                response.addHeader(WWW_AUTHENTICATE, "Basic error='invalid_literal' error-desc='Invalid username:password string' realm='https://account.companieshouse.gov.uk'");
```
Comment (emead) [inline src/main/java/uk/gov/companieshouse/accounts/user/interceptor/Oauth2AuthorizationInterceptor.java:71]: The realm should be `https://identity.company-information.service.gov.uk`

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by jackthurkettle — APPROVED** [review]

**Pull request merged**